src/librustc_trans/back/linker.rs

```diff
@@ -115,30 +115,32 @@ impl<'a> Linker for GnuLinker<'a> {
     }
 
     fn gc_sections(&mut self, is_dylib: bool) {
-        // The dead_strip option to the linker specifies that functions and data
-        // unreachable by the entry point will be removed. This is quite useful
-        // with Rust's compilation model of compiling libraries at a time into
-        // one object file. For example, this brings hello world from 1.7MB to
-        // 458K.
-        //
-        // Note that this is done for both executables and dynamic libraries. We
-        // won't get much benefit from dylibs because LLVM will have already
-        // stripped away as much as it could. This has not been seen to impact
-        // link times negatively.
-        //
-        // -dead_strip can't be part of the pre_link_args because it's also used
-        // for partial linking when using multiple codegen units (-r).  So we
-        // insert it here.
-        if self.sess.target.target.options.is_like_osx {
-            self.cmd.arg("-Wl,-dead_strip");
-
-        // If we're building a dylib, we don't use --gc-sections because LLVM
-        // has already done the best it can do, and we also don't want to
-        // eliminate the metadata. If we're building an executable, however,
-        // --gc-sections drops the size of hello world from 1.8MB to 597K, a 67%
-        // reduction.
-        } else if !is_dylib {
-            self.cmd.arg("-Wl,--gc-sections");
+        if !self.sess.opts.cg.no_gc_sections {
```

Comment:
(but also, I am the kind of person that would put the conditional check around the call-site to `linker.gc_sections(..)`; this change is making this function act more like `fn maybe_gc_sections(..)`)